### PR TITLE
CML workflow on push & PR

### PR DESCRIPTION
**Pull request by casperdcl**
## Currently:

TL;DR: run on `push` and `pr` (incl. fetching `main` for diff)

- on commit to `main`: [compare to previous commit, report as a commit-comment](https://github.com/iterative/example-get-started/commit/7fe0949204cdef55886ef23b4ed697a311b500b5#comments)
- on PR: [compare to `main`, report as a PR-comment](https://github.com/iterative/example-get-started/pull/31#issuecomment-1281051987)
- on commit to non-`main` (incl. updating a PR): treat as PR
- unwanted side-effect of ignoring failures on non-pr-featurebranch-push

## Previously in [c3b04d0](https://github.com/iterative/example-get-started/pull/31/commits/c3b04d0af303c38801b3840e4e27042c17b7db33):

TL;DR: restrict `on.push.branches: [main]`, and add `on: pr` (incl. fetching `main` for diff)

- on commit to `main`: [compare to previous commit, report as a commit-comment](https://github.com/iterative/example-get-started/commit/7fe0949204cdef55886ef23b4ed697a311b500b5#comments)
- on PR: [compare to `main`, report as a PR-comment](https://github.com/iterative/example-get-started/pull/31#issuecomment-1281051987)
- ~on both (push new commit to PR): do both~ inapplicable (`main` can't PR to itself)
- unwanted side-effect of not running on non-pr-featurebranch-push

<details><summary>Outdated debugging</summary>

WiP. Doing battle with GHA. Notoriously verbose-yet-somehow-not-informative docs.

```yaml
on: [push, pull_request]
concurrency:
  group: ${{ github.workflow }}-${{ github.event.pull_request.head.sha || github.sha }}
  cancel-in-progress: ${{ github.event_name == 'push' }}
```

this :point_up: still doesn't solve the problem, right? I'm trying to cancel `push` jobs if there's a corresponding `pull_request` job.

- does `cancel-in-progress` accept dynamic values? It's not [documented](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#concurrency)
  + if so, what happens if the `push` job runs while the `pull_request` job is queued? The `cancel-in-progress` will never be triggered and duplicate runs will happen, right?
  + alternatively if `cancel-in-progress` takes the most recent value assigned to it, again there will be cases where it's not triggered
  + based on [example-get-started#31@b30b0a0](https://github.com/iterative/example-get-started/pull/31/commits/b30b0a0e785dfebfd4fd5debbbeca19de43d0677), it looks like it's not triggered (both the commit comment and the PR comment were created/updated)
- alternatively, `cancel-in-progress: true` will trigger cancellation of either the `push` or the `pull_request` but in an unpredictable manner... sometimes meaning a commit-comment, sometimes meaning a pr-comment
- ~worse, [example-get-started#31@3a4e255](https://github.com/iterative/example-get-started/pull/31/commits/3a4e255e18f929f574d8cec58d92c7009c4f590c) indicates `cancel-in-progress: true` is just ignored? Both jobs run? Does it need to be committed to `main` to work?~

</details>

### CODE REVIEW

File: .github/workflows/cml.yaml [hunk 1]
```diff
@@ -14,10 +14,10 @@ jobs:
           REPO_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         run: |
           cml ci
-          if [ $GITHUB_REF = refs/heads/main ]; then
-            PREVIOUS_REF=HEAD~1
-          else
+          if test ${{ github.event_name }} = pull_request; then
             PREVIOUS_REF=main
```
Comment (casperdcl) [inline .github/workflows/cml.yaml]:
```suggestion
            PREVIOUS_REF=main
            git fetch --depth=1 origin $PREVIOUS_REF:$PREVIOUS_REF
```

File: .github/workflows/cml.yaml [hunk 2]
```diff
@@ -1,5 +1,8 @@
 name: CML Report
-on: push
+on: [push, pull_request]
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.ref }}
```
Comment (dacbd) [inline .github/workflows/cml.yaml]:
```suggestion
  group: ${{ github.workflow }}-${{ github.head_ref || github.ref_name }}
```

Comment (casperdcl) [reply .github/workflows/cml.yaml]: Actually not sure this is right, as it relies on names rather than SHAs (so susceptible to race conditions).

[`github` context](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) | Name | SHA
-|-|-
source | `head_ref`[^pr], `ref_name`[^diff] | `event.pull_request.head.sha`[^pr], `sha`[^diff]
target | `base_ref`[^pr] | -

[^diff]: different values for `push` & for `pr`
[^pr]: only defined for `pr`

```suggestion
  group: ${{ github.workflow }}-${{ github.event.pull_request.head.sha || github.sha }}
```

Use case here: if a PR is updated, we do *not* want to cancel the running job. They could be distinct experiments in one PR. We only want to cancel `push` if there's a corresponding `pr` (i.e. no duplicate runs).

File: .github/workflows/cml.yaml [hunk 3]
```diff
@@ -9,15 +12,17 @@ jobs:
       - uses: actions/checkout@v3
         with:
           fetch-depth: 2
+          ref: ${{ github.head_ref || github.ref }}
```
Comment (casperdcl) [inline .github/workflows/cml.yaml]:
```suggestion
          ref: ${{ github.event.pull_request.head.sha || github.sha }}
```

File: .github/workflows/cml.yaml [hunk 4]
```diff
@@ -1,5 +1,8 @@
 name: CML Report
-on: push
+on: [push, pull_request]
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.ref_name }}
```
Comment (casperdcl) [inline .github/workflows/cml.yaml]:
```suggestion
  group: ${{ github.workflow }}-${{ github.event.pull_request.head.sha || github.sha }}
```

File: .github/workflows/cml.yaml [hunk 5]
```diff
@@ -9,15 +12,17 @@ jobs:
       - uses: actions/checkout@v3
         with:
           fetch-depth: 2
+          ref: ${{ github.event.pull_request.head.sha || github.sha }}
```
Comment (casperdcl) [inline .github/workflows/cml.yaml]:
```suggestion
          ref: ${{ github.event.pull_request.head.sha }}
```

File: .github/workflows/cml.yaml [hunk 6]
```diff
@@ -1,5 +1,8 @@
 name: CML Report
-on: push
+on: [push, pull_request]
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.head.sha || github.sha }}
+  cancel-in-progress: true
 jobs:
   run:
     runs-on: [ubuntu-latest]
```
Comment (casperdcl) [inline .github/workflows/cml.yaml]:
```suggestion
    continue-on-error: ${{ github.ref_name != 'main' }}
    runs-on: [ubuntu-latest]
```

File: .github/workflows/cml.yaml [hunk 7]
```diff
@@ -1,23 +1,29 @@
...
         with:
           fetch-depth: 2
+          ref: ${{ github.event.pull_request.head.sha }}
       - name: Generate metrics report
         env:
           REPO_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         run: |
           cml ci
-          if [ $GITHUB_REF = refs/heads/main ]; then
+          if test $GITHUB_REF = refs/heads/main; then
```
Comment (casperdcl) [inline .github/workflows/cml.yaml]:
```suggestion
          if test ${{ github.ref_name }} = main; then
```

File: .github/workflows/cml.yaml [hunk 8]
```diff
@@ -1,23 +1,29 @@
 name: CML Report
-on: push
+on: [push, pull_request]
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.head.sha || github.sha }}
+  cancel-in-progress: true
 jobs:
   run:
+    continue-on-error: ${{ github.ref_name != 'main' }}
     runs-on: [ubuntu-latest]
```
Comment (casperdcl) [inline .github/workflows/cml.yaml]:
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/cml.yaml [hunk 9]
```diff
@@ -1,23 +1,29 @@
...
         with:
           fetch-depth: 2
+          ref: ${{ github.event.pull_request.head.sha }}
       - name: Generate metrics report
         env:
           REPO_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         run: |
           cml ci
-          if [ $GITHUB_REF = refs/heads/main ]; then
+          if test ${{ github.ref_name }} = main; then
```
Comment (0x2b3bfa0) [inline .github/workflows/cml.yaml]:
```suggestion
          if test $GITHUB_REF_NAME == main; then
```

Alternative spelling ™️

**Review by 0x2b3bfa0 — APPROVED** [review]

**Comment by 0x2b3bfa0** [comment]
What's the point of using commit hashes for concurrency? 🤔

**Comment by skshetry** [comment]
This repo is autogenerated using https://github.com/iterative/example-repos-dev.

File: .github/workflows/cml.yaml [hunk 10]
```diff
@@ -1,23 +1,29 @@
 name: CML Report
-on: push
+on: [push, pull_request]
+concurrency:
```
Comment (shcheklein) [inline .github/workflows/cml.yaml]: Q: what does it give us? Can we keep it simple(r) for this repo (even if we hit some edge cases)?

File: .github/workflows/cml.yaml [hunk 11]
```diff
@@ -1,23 +1,29 @@
...
+  group: ${{ github.workflow }}-${{ github.event.pull_request.head.sha || github.sha }}
+  cancel-in-progress: true
 jobs:
   run:
-    runs-on: [ubuntu-latest]
+    continue-on-error: ${{ github.ref_name != 'main' }}
```
Comment (shcheklein) [inline .github/workflows/cml.yaml]: Q: what does it give us - same question as above pretty much ...

Comment (dacbd) [reply .github/workflows/cml.yaml]: Agreed, this is essentially saying even if this job fails report it as successful unless it runs against the main branch.

In theory, it could be nice for a user as it would reduce notifications to their email, but there are better ways to solve that problem IMO. This not how you normally see this feature used and I don't think it should be in a getting started type example.

Comment (casperdcl) [reply .github/workflows/cml.yaml]: no, it's saying even if this job is auto-cancelled (because it is a duplicate `push` job that has a corresponding `pull_request` job), don't treat the auto-cancellation as an error.

Without this `continue-on-error` (ab)use, you [get an error](https://github.com/iterative/example-get-started/actions/runs/3301197176).

File: .github/workflows/cml.yaml (hunk 10 above)
Comment (casperdcl) [reply .github/workflows/cml.yaml]: - right now, we can't keep it simpler for the desired behaviour
- if we're willing to drop/change the desired behaviour, we can simplify this
- this PR currently demonstrates how hard it is to use GHA/CML to do what should be extremely simple. I intend it as something we should write a technical article about. I do not think we should merge this here nor into https://github.com/iterative/example-repos-dev. It is wayyy to complex.

Comment (casperdcl) [reply .github/workflows/cml.yaml]: WDYT about the ["**Previously in `c3b04d0`**" version](https://github.com/iterative/example-get-started/pull/31#issue-1411746987)?

File: .github/workflows/cml.yaml (hunk 11 above)
Comment (casperdcl) [reply .github/workflows/cml.yaml]: Mind you; unwanted side-effect of `continue-on-error: ${{ github.ref_name != 'main' }}`: silently ignores non-pr-branch-push jobs that result in failures.

File: .github/workflows/cml.yaml (hunk 10 above)
Comment (shcheklein) [reply .github/workflows/cml.yaml]: I think let's step back and do the simplest one - PR, no main/master, no ifs, no cancellation, etc. We are going a bit too far with this tbh.

File: .github/workflows/cml.yaml (hunk 11 above)
Comment (casperdcl) [reply .github/workflows/cml.yaml]: also I guess this is better:

```suggestion
    continue-on-error: ${{ github.ref_name != github.event.repository.default_branch }}
```

(seen in https://github.com/iterative/link-check/pull/18#discussion_r933000447)

**Comment by 0x2b3bfa0** [comment]
@casperdcl, please open a pull request from https://github.com/iterative/example-repos-dev/compare/cml-push-and-pr?expand=1 if necessary.
